File: model/src/main/scala/ru/mipt/acsl/decode/model/types/TypeMeasure.scala
package ru.mipt.acsl.decode.model.types

import ru.mipt.acsl.decode.model.naming.{ElementName, Namespace}
import ru.mipt.acsl.decode.model.proxy.MaybeProxy
import ru.mipt.acsl.decode.model.registry.Measure

/**
  * @author Artem Shein
  */
trait TypeMeasure extends DecodeType {

  def typeProxy: MaybeProxy.TypeProxy

  def t: DecodeType = typeProxy.obj

  def measureProxy: Option[MaybeProxy.Measure]

  def measure: Option[Measure] = measureProxy.map(_.obj)

  override def namespace: Namespace = t.namespace

  override def namespace_=(ns: Namespace): Unit = t.namespace = ns

  override def alias: Option[Alias.Type] = t.alias

  override def systemName: String = t.systemName

  override def typeParameters: Seq[ElementName] = t.typeParameters

  override def toString: String =
    s"${this.getClass}{alias = $alias, namespace = $namespace, t = $t, measure = $measure}"

}

object TypeMeasure {

  private case class TypeMeasureImpl(typeProxy: MaybeProxy.TypeProxy, measureProxy: Option[MaybeProxy.Measure])
    extends TypeMeasure

  def apply(typeProxy: MaybeProxy.TypeProxy, unitProxy: Option[MaybeProxy.Measure]): TypeMeasure =
    TypeMeasureImpl(typeProxy, unitProxy)
}

File: model/src/main/java/ru/mipt/acsl/decode/model/types/TypeMeasure.java
package ru.mipt.acsl.decode.model.types;

import org.jetbrains.annotations.Nullable;
import ru.mipt.acsl.decode.model.Referenceable;
import ru.mipt.acsl.decode.model.ReferenceableVisitor;
import ru.mipt.acsl.decode.model.naming.ElementName;
import ru.mipt.acsl.decode.model.naming.Namespace;
import ru.mipt.acsl.decode.model.proxy.MaybeProxyMeasure;
import ru.mipt.acsl.decode.model.proxy.MaybeTypeProxy;
import ru.mipt.acsl.decode.model.registry.Measure;

import java.util.List;
import java.util.Optional;

/**
 * Created by metadeus on 07.06.16.
 */
public interface TypeMeasure extends Referenceable {

    static TypeMeasure newInstance(MaybeTypeProxy typeProxy, @Nullable MaybeProxyMeasure measureProxy) {
        return new TypeMeasureImpl(typeProxy, measureProxy);
    }

    MaybeTypeProxy typeProxy();

    default DecodeType t() {
        return typeProxy().obj();
    }

    Optional<MaybeProxyMeasure> measureProxy();

    default Optional<Measure> measure() {
        return measureProxy().map(MaybeProxyMeasure::obj);
    }

    default Namespace namespace() {
        return t().namespace();
    }

    default Optional<Alias> alias() {
        return t().alias();
    }

    default String systemName() {
        return t().systemName();
    }

    default List<ElementName> typeParameters() {
        return t().typeParameters();
    }

    @Override
    default <T> T accept(ReferenceableVisitor<T> visitor) {
        return visitor.visit(this);
    }
}
